packages/puppeteer-core/src/bidi/Input.ts

```diff
@@ -613,79 +615,86 @@ export class BidiMouse extends Mouse {
 /**
  * @internal
  */
-export class BidiTouchscreen extends Touchscreen {
+class BidiTouch implements TouchHandle {
+  #started = false;
+  #x: number;
+  #y: number;
+  #bidiId: string;
   #page: BidiPage;
+  #touchScreen: BidiTouchscreen;
+  #properties: Bidi.Input.PointerCommonProperties;
 
-  constructor(page: BidiPage) {
-    super();
+  constructor(
+    page: BidiPage,
+    touchScreen: BidiTouchscreen,
+    public id: number,
+    x: number,
+    y: number,
+    properties: Bidi.Input.PointerCommonProperties,
+  ) {
     this.#page = page;
+    this.#touchScreen = touchScreen;
+    this.#x = Math.round(x);
+    this.#y = Math.round(y);
+    this.#properties = properties;
+    this.#bidiId = `${InputId.Finger}_${id}`;
   }
 
-  override async touchStart(
-    x: number,
-    y: number,
-    options: BidiTouchMoveOptions = {},
-  ): Promise<void> {
+  async start(): Promise<void> {
+    if (this.#started) {
+      throw new TouchError('Touch has already started');
+    }
     await this.#page.mainFrame().browsingContext.performActions([
       {
         type: SourceActionsType.Pointer,
-        id: InputId.Finger,
+        id: this.#bidiId,
         parameters: {
           pointerType: Bidi.Input.PointerType.Touch,
         },
         actions: [
           {
             type: ActionType.PointerMove,
-            x: Math.round(x),
-            y: Math.round(y),
-            origin: options.origin,
```

Comment:
I think we need to keep the origin.

Reply:
Yes that is something I didn't understand. [This parameter](https://github.com/puppeteer/puppeteer/blob/b60067735da000200b1b151a0cfb5ace5bf75738/packages/puppeteer-core/src/bidi/Input.ts#L659) and [this one](https://github.com/puppeteer/puppeteer/blob/b60067735da000200b1b151a0cfb5ace5bf75738/packages/puppeteer-core/src/bidi/Input.ts#L627) are not part of the public api, and they're not tested at all. Only the Bidi touchscreen has them, and not the Cdp one. How are they ever used?

Reply:
I put the `BidiTouchMoveOptions` parameters back, but they're still untested.